fix!: set out hash root in checkpoint header and validate on L1

LeilaWang opened the pull request:
Add an "epoch" `outHash` to the checkpoint header, which is the root of an append-only sha merkle tree that contains the checkpoint's out hash and the out hashes of its previous checkpoints in the same epoch - the out hash of the first checkpoint is at index 0, the second at index 1, and so on. The `outHash` of the last checkpoint in an epoch will match the final epoch `outHash` (the `outHash` containing all tx out hashes in an epoch), which will be inserted to the `Outbox` on L1.

iAmMichaelConnor commented on noir-projects/noir-protocol-circuits/crates/rollup-lib/src/checkpoint_root/components/checkpoint_rollup_public_inputs_composer.nr:
```diff
@@ -90,7 +97,15 @@ impl<let NumBlobs: u32> CheckpointRollupPublicInputsComposer<NumBlobs> {
 
-    fn create_checkpoint_header(self) -> CheckpointHeader {
+    fn accumulate_checkpoint_out_hash(self, checkpoint_out_hash: Field) -> AppendOnlyTreeSnapshot {
+        append_leaf_to_snapshot_sha::<OUT_HASH_TREE_HEIGHT>(
```
Not a question, just a note to myself for future:

6 sha256 hashes to append this leaf. That's a shame, but I agree it's the simplest approach.

The only other way would be `hash(...hash(hash(out_hash_1), out_hash_2)..., out_hash_n)` (which we chatted about the other day), but as you pointed out that makes it very cumbersome to _consume_ a message on L1: a user would have to say "I'm consuming a message from epoch 1234, which had 5 checkpoints, so I need to do 5 layers of hashing on L1 to consume this message". Too cumbersome at this late stage. Besides, these extra thousands of constraints are a small fraction of this circuit's constraints.

iAmMichaelConnor commented on noir-projects/noir-protocol-circuits/crates/types/src/abis/checkpoint_header.nr:
```diff
@@ -63,30 +66,31 @@ impl CheckpointHeader {
 
         for i in 0..20 {
-            bytes[i + 168] = coinbase_bytes[i];
+            bytes[i + 200] = coinbase_bytes[i];
```
Should we do `offset += thing.len()`, `bytes[i + offset]` here, instead of the hard-coded numbers?

iAmMichaelConnor commented on noir-projects/noir-protocol-circuits/crates/types/src/constants.nr:
```diff
@@ -162,6 +164,8 @@ pub global GENESIS_BLOCK_HEADER_HASH: Field =
     0x0dc66232e4edff2923217f4fdf80ecc23ed4bd047e743df85a6827878bf6c4d6;
 pub global GENESIS_ARCHIVE_ROOT: Field =
     0x1f8c805403d88ee809d3cb3e52eeba0a104f69968cdb844a9cbdf1e9e00b3a95;
+pub global EMPTY_EPOCH_OUT_HASH: Field =
```
Should we call it `...OUT_HASH_TREE_ROOT`, for consistency with the other constants which also refer to it as the `OUT_HASH_TREE`?
Edit: Oh, looks like we also refer to this "out hash tree root" as the "out hash" in the L1 contracts.

iAmMichaelConnor replied:
Also, is there a test to check that this constant does indeed represent an empty tree?

LeilaWang replied:
The test is in `out_hash_tests.nr`, I'll move it here!

iAmMichaelConnor commented on noir-projects/noir-protocol-circuits/crates/types/src/constants.nr:
```diff
@@ -1328,4 +1333,9 @@ mod test {
+    fn validate_out_hash_tree_height() {
+        let max_out_hash_leaves = 1 << OUT_HASH_TREE_HEIGHT;
+        assert(max_out_hash_leaves >= AZTEC_MAX_EPOCH_DURATION);
+    }
```
Perhaps we could also include a check like:

```
assert(AZTEC_MAX_EPOCH_DURATION * 2 > max_out_hash_leaves, "You can get away with making your out_hash tree smaller, by the way");
```

to help us realise that the tree is too big, if we ever change the number of checkpoints in an epoch

LeilaWang replied:
Good idea!

iAmMichaelConnor commented on l1-contracts/src/core/libraries/ConstantsGen.sol:
```diff
@@ -21,6 +21,8 @@ library Constants {
   uint256 internal constant AZTEC_MAX_EPOCH_DURATION = 48;
   uint256 internal constant GENESIS_ARCHIVE_ROOT =
     14_269_942_583_723_164_841_365_114_274_712_143_548_835_546_030_057_296_325_580_016_468_921_911_294_613;
+  uint256 internal constant EMPTY_EPOCH_OUT_HASH =
+    425_901_435_741_743_301_163_237_390_758_864_716_627_999_612_705_062_243_139_499_504_804_273_333_597;
```
Aside (not related to this PR): do we have some tests that ensure all these generated constants actually match the original `constants.nr` constants?
I'm like "I guess this is correct, because I trust the auto-generation code".

LeilaWang replied:
That's a good point! For this particular variable there's a ts test that tests it. But there could be problem with other values. I'll think about how to test it after the audit!

iAmMichaelConnor reviewed (approved):
Circuits look good. Just a couple of minor q's/comments.
Lasse will need to sign off on the L1 changes. They look fine to me, but I'm not familiar enough with them to be comfortable approving them.

spalladino commented on yarn-project/end-to-end/src/e2e_multi_eoa.test.ts:
```diff
@@ -126,35 +126,39 @@ describe('e2e_multi_eoa', () => {
-            blockedTxs.push(txHash);
-            logger.warn(`Blocking tx from sender ${signerAddress.toString()} with hash ${txHash}`);
-            return txHash;
+      // Get unique clients - they may or may not be the same object
+      const uniqueClients = [...new Set(l1Utils.map(u => u.client))];
+      const originalSendRawTransactions = new Map(uniqueClients.map(client => [client, client.sendRawTransaction]));
```
For my own understanding: is this change needed because we're creating new viem clients that we were not creating before?

LeilaWang replied:
Yeah. This test was failing because the call was not mocked for the second address, and I realized that there were actually 2 different clients here.

spalladino commented on yarn-project/stdlib/src/checkpoint/checkpoint.ts:
```diff
@@ -66,6 +67,14 @@ export class Checkpoint {
     return this.header.hash();
   }
 
+  // Returns the out hash computed from all l2-to-l1 messages in this checkpoint.
+  // Note: This value is different from the out hash in the header, which is the **accumulated** out hash over all
+  // checkpoints up to and including this one in the epoch.
```
I can see a lot of potential for confusion if we have a `Checkpoint.outHash` and a `CheckpointHeader.outHash` that represent different things. Can we rename one of them? Maybe the one in the header one to `epochOutHash`?

LeilaWang replied:
Will change it to `epochOutHash`. It's much better. Thanks for the suggestion!

spalladino commented on yarn-project/stdlib/src/messaging/out_hash.ts:
```diff
@@ -1,46 +1,50 @@
+import { computeBalancedShaRoot, computeUnbalancedShaRoot, computeWonkyShaRoot } from '@aztec/foundation/trees';
 
+// Computes the out hash of a tx. It's the root of the unbalanced tree of all the L2-to-L1 messages emitted by the tx.
+// Must match the implementation in `tx_base_public_inputs_composer.nr > compute_kernel_out_hash()`.
```
Nit: Let's use jsdoc syntax for this kind of comments so they come up in intelisense. No need to change now, just for the future.

spalladino commented on yarn-project/validator-client/src/block_proposal_handler.ts:
```diff
@@ -196,14 +204,44 @@ export class BlockProposalHandler {
     // Try re-executing the transactions in the proposal if needed
     let reexecutionResult;
     if (shouldReexecute) {
+      // Compute the previous checkpoint out hashes for the epoch.
+      const epoch = getEpochAtSlot(slotNumber, this.epochCache.getL1Constants());
+      const previousBlocks = (await this.blockSource.getBlocksForEpoch(epoch))
+        .filter(b => b.number < blockNumber)
+        .sort((a, b) => a.number - b.number);
+      const previousCheckpointOutHashes = previousBlocks.map(b =>
+        computeCheckpointOutHash([b.body.txEffects.map(tx => tx.l2ToL1Msgs)]),
+      );
```
Shouldn't we be grouping blocks by checkpoint here before calling `computeCheckpointOutHash`? Is this assuming one block per checkpoint? If so, I can change it in #19240, where I'm tweaking the block proposal handler to work with full checkpoints. Can you add a TODO so we don't forget the places where we still assume one block per checkpoint?

spalladino replied:
Also, is there a way we can avoid recomputing this every time we need to reexecute a block? I'd like to avoid having to load every tx in an epoch just to compute these `outHash`es. No need to implement now, but let's keep it for a future PR.

LeilaWang replied:
Yes it should be an array of blocks for each checkpoint. Actually I should've called `getCheckpointOutHash`.  This was first written before checkpoint even existed in the ts land 😛 
I'll add a todo for using a more efficient way to retrieve the out hashes!

spalladino commented on yarn-project/slasher/src/watchers/epoch_prune_watcher.ts:
```diff
@@ -119,21 +120,31 @@ export class EpochPruneWatcher extends (EventEmitter as new () => WatcherEmitter
+        previousCheckpointOutHashes = [
+          ...previousCheckpointOutHashes,
+          computeCheckpointOutHash([block.body.txEffects.map(tx => tx.l2ToL1Msgs)]),
+        ];
```
Let's also add a comment here to not assume one block per checkpoint

spalladino commented on yarn-project/foundation/src/trees/unbalanced_merkle_tree_root.ts:
```diff
@@ -105,7 +100,11 @@ export async function computeUnbalancedMerkleTreeRootAsync(
 }
 
-export function computeCompressedUnbalancedMerkleTreeRoot(
+// A **wonky** tree is a "compressed" unbalanced Merkle tree.
+// It is constructed in the same way as an unbalanced tree: by first creating the largest possible left subtree, with
+// the remaining leaves forming a right subtree that follows the same process recursively.
+// During construction, leaves equal to `valueToCompress` are skipped (compressed) and do not contribute to the tree.
```
At some point it'd be good to have Claude reorganize all the trees we have spread over the codebase (here and in the merkle-tree package) and produce a readme explaining all the flavors we have.

spalladino commented on yarn-project/prover-client/src/block-factory/light.ts:
```diff
@@ -87,6 +92,10 @@ export class LightweightBlockFactory implements IBlockFactory {
     const newArchive = await getTreeSnapshot(MerkleTreeId.ARCHIVE, this.db);
 
+    const blockOutHash = computeBlockOutHash(txs.map(tx => tx.txEffect.l2ToL1Msgs));
+    // There's only one block per checkpoint, so the checkpoint out hash equals the block out hash.
```
TODO here as well!

spalladino commented on yarn-project/prover-client/src/orchestrator/checkpoint-proving-state.ts:
```diff
@@ -38,6 +40,11 @@ import { accumulateBlobs, buildBlobHints, toProofData } from './block-building-h
 import { BlockProvingState, type ProofState } from './block-proving-state.js';
 import type { EpochProvingState } from './epoch-proving-state.js';
 
+type OutHashHints = {
```
```suggestion
type OutHashHint = {
```
IIUC it's a single hint for a single value, right?

spalladino commented on yarn-project/prover-client/src/orchestrator/orchestrator.ts:
```diff
@@ -352,7 +361,8 @@ export class ProvingOrchestrator implements EpochProver {
     provingState.setEndSpongeBlob(spongeBlobState);
 
-    // Txs have been added to the block. Now try to accumulate the blobs as far as we can:
+    // Txs have been added to the block. Now try to accumulate the out hashes and blobs as far as we can:
+    await this.provingState.accumulateCheckpointOutHashes();
```
Shouldn't we wait until the checkpoint is complete before trying to accumulate?

LeilaWang replied:
It doesn't have to be complete. Once all txs in the previous checkpoints are added, we will be able to fetch the l2-to-l1 msgs to build the tree and construct the private inputs for proving the current checkpoint.

spalladino commented on yarn-project/prover-client/src/orchestrator/orchestrator.ts:
```diff
@@ -255,7 +263,8 @@ export class ProvingOrchestrator implements EpochProver {
       blockProvingState.setEndSpongeBlob(endSpongeBlob);
 
-      // And also try to accumulate the blobs as far as we can:
+      // Try to accumulate the out hashes and blobs as far as we can:
+      await this.provingState.accumulateCheckpointOutHashes();
```
Same on waiting until checkpoint is complete

spalladino reviewed (commented):
Ts-land looks good! I left a couple of questions, but my main points are:
- Let's leave a TODO in the code everywhere we assume one block per checkpoint, so we have an easier time changing that.
- Let's try not to use the same name `outHash` for different things (as in the checkpoint out hash and the checkpoint header out hash)

LHerskind commented on l1-contracts/src/core/libraries/compressed-data/CheckpointLog.sol:
```diff
@@ -20,6 +20,7 @@ struct CheckpointLog {
   bytes32 archive;
   bytes32 headerHash;
   bytes32 blobCommitmentsHash;
+  bytes32 outHash;
```
![Image](https://github.com/user-attachments/assets/47070751-202d-48f7-84a5-3c990c9a8493)

LHerskind commented on l1-contracts/src/core/libraries/rollup/EpochProofLib.sol:
```diff
@@ -314,6 +317,9 @@ library EpochProofLib {
 
     ValidatorSelectionLib.verifyAttestations(slot, epoch, _attestations, checkpointLog.payloadDigest);
+
+    // Verify that the out hash matches the stored value
```
I think it might benefit us to mention that it is a component of the payloadDigest used to validate signatures against right above. At least that was the first thing I were looking into here to make sure that the value was included in the attestations signatures 👍.

Not sure how to word that nicely though, so maybe not much here really.

LHerskind commented on l1-contracts/src/core/libraries/rollup/ProposedHeaderLib.sol:
```diff
@@ -54,6 +55,7 @@ library ProposedHeaderLib {
         _header.blockHeadersHash,
         _header.blobsHash,
         _header.inHash,
+        _header.outHash,
```
Related to above as this hash makes its way into payloadDigest

LHerskind reviewed (commented):
Looked at the L1, makes sense to me. Mainly just that it could benefit from a comment to make it very clear that it is part of the signed messaged 🫡

LeilaWang committed:
fix!: set out hash root in checkpoint header and validate on L1

Add an "epoch" `outHash` to the checkpoint header, which is the root of an append-only sha merkle tree that contains the checkpoint's out hash and the out hashes of its previous checkpoints in the same epoch - the out hash of the first checkpoint is at index 0, the second at index 1, and so on. The `outHash` of the last checkpoint in an epoch will match the final epoch `outHash` (the `outHash` containing all tx out hashes in an epoch), which will be inserted to the `Outbox` on L1.
